### PR TITLE
idGenerator request object (#62)

### PR DESCRIPTION
**Pull request by Legolaszstudio**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

#### What did I do?

Fix for #62 
Added the request object to the idGenerator function for custom scenarios like:
```js
idGenerator: (request) => {
     if (request.session.returningVisitor) return `returningVisitor-${uid(24)}`
     else return uid(24)
}
```

### CODE REVIEW

File: types/types.d.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ interface SessionData extends ExpressSessionData {
   touch(): void;
 
   /** Regenerates the session by generating a new `sessionId`. */
-  regenerate(): void;
```
Comment (Legolaszstudio) [inline types/types.d.ts:30]: Should this be forced like this or should I change to a nullable parameter?

File: types/types.d.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ interface SessionData extends ExpressSessionData {
...
+  regenerate(request: Fastify.FastifyRequest): void;
```
Comment (Legolaszstudio) [inline types/types.d.ts:30]: Should I change to a nullable or keep it like this?

**Comment by Legolaszstudio** [comment]
I wrote a very simple test case, and in my opinion, there isn't really anything else to test.
In case there is an edge case or something like that, it won't really cover that....

File: lib/session.js [hunk 3]
```diff
@@ -55,8 +55,8 @@ module.exports = class Session {
     return cookieSignature.sign(this.sessionId, this[secretKey])
   }
 
-  static restore (idGenerator, cookieOpts, secret, prevSession) {
-    const restoredSession = new Session(idGenerator, cookieOpts, secret, prevSession)
+  static restore (request, idGenerator, cookieOpts, secret, prevSession) {
```
Comment (mcollina) [inline lib/session.js:58]: Is this change tested?

File: types/types.d.ts (hunk 2 above)
Comment (mcollina) [reply types/types.d.ts:30]: cc @fastify/typescript

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you please add a test for the types? We use tsd.

**Review by darkgl0w — CHANGES_REQUESTED** [review]
Some changes and types unit tests are missing.

File: test/session.test.js [hunk 4]
```diff
@@ -345,3 +345,58 @@ test('should update the expires property of the session using Session#touch() ev
...
 })
+
+test('should use custom sessionId generator if available (with request)', async (t) => {
+  const fastify = Fastify()
+  fastify.register(fastifyCookie)
+  fastify.register(fastifySession, {
+    secret: 'cNaoPYAwF60HZJzkcNaoPYAwF60HZJzk',
+    cookie: { secure: false, maxAge: 10000 },
+    idGenerator: (request) => {
+      if (request?.session.returningVisitor) {
```
Comment (darkgl0w) [inline test/session.test.js:356]: Optional chaining is not supported by Node.js 10

File: types/types.d.ts (hunk 2 above)
Comment (Legolaszstudio) [reply types/types.d.ts:30]: Ok, let's make it optional, so we don't break anything.
But make sure that it is clearly documented.

File: test/session.test.js (hunk 4 above)
Comment (Legolaszstudio) [reply test/session.test.js]: Fixed in
https://github.com/fastify/session/pull/64/commits/99614dad9a9cd78d6bc0daec75ba64d4723ebff9

File: types/types.d.ts [hunk 5]
```diff
@@ -101,7 +101,7 @@ declare namespace FastifySessionPlugin {
     rolling?: boolean;
 
     /** Function used to generate new session IDs. Defaults to uid(24). */
-    idGenerator?: () => string;
+    idGenerator?: (request: Fastify.FastifyRequest) => string;
```
[the remark this answers is not in the document]
Comment (Legolaszstudio) [reply types/types.d.ts]: It is optional indeed, my mistake

File: lib/session.js (hunk 3 above)
Comment (Legolaszstudio) [reply lib/session.js:58]: Restore is only referenced once; To be exact in fastifySession line 80 (https://github.com/fastify/session/pull/64/files#diff-c327147bd50b6dd851469add08b782e4ede4da6003ddbab99f46e50ff220d84bR80-R84)

I don't see how this could be tested? I only added a request parameter to it which will be called inside Session's constructor which is tested in theory. Nevertheless I will add a rolling false custom id generator test case

**Comment by Legolaszstudio** [comment]
Fixed everything mentioned, except:
> Could you please add a test for the types? We use tsd.

How do I test the types and what should I test on them?

File: types/types.d.ts [hunk 6]
```diff
@@ -26,8 +26,11 @@ interface SessionData extends ExpressSessionData {
   /** Updates the `expires` property of the session. */
   touch(): void;
 
-  /** Regenerates the session by generating a new `sessionId`. */
-  regenerate(): void;
+  /**
+   * Regenerates the session by generating a new `sessionId`.
+   * @param request Optional, but REQUIRED if idGenerator uses it
+   */
+  regenerate(request?: Fastify.FastifyRequest): void;
```
Comment (darkgl0w) [inline types/types.d.ts:33]:
```suggestion
  /**
   * Regenerates the session by generating a new `sessionId`.
   * @param request optional, but __REQUIRED__ if the `idGenerator` function used to generate new session IDs uses it.
   */
  regenerate(request?: Fastify.FastifyRequest): void;
```

Comment (darkgl0w) [reply types/types.d.ts:33]: One other way to type the `regenerate` method we can explore would be something like this:
```ts
// With this we can get the parameters of a function type in a tuple
type Parameters<T> = T extends (...args: infer P) => any ? P : never;

// and by typing regenerate like this it will accept the same arguments (and types) that `idGenerator` accepts
// NB: if we add another argument to `idGenerator` it will be added to `regenerate` too.
regenerate(...args: Parameters<FastifySessionPlugin.Options['idGenerator']>): void;
```

cc @fastify/typescript WDYT ?

**Review by darkgl0w — CHANGES_REQUESTED** [review]
A few nitpicks and typescript definitions tests are missing.

Take a look at the `types/types.test-d.ts` to see how we test our typescript definitions. ^^
You need to add a case where you register the plugin with an `idGenerator` function that takes a `request` argument and uses it to generate a `sessionId`, and by extension a `regenerate` use case that takes a `request` argument too.

File: README.md [hunk 7]
```diff
@@ -104,9 +111,15 @@ Allows to destroy the session in the store
 
 Updates the `expires` property of the session.
 
-#### Session#regenerate()
+#### Session#regenerate(request)
 
-Regenerates the session by generating a new `sessionId`.
+Regenerates the session by generating a new `sessionId`. Don't forget to pass the fastify.request object if your id generator uses it.
```
Comment (Eomm) [inline README.md:116]:
```suggestion
Regenerates the session by generating a new `sessionId`. Don't forget to pass the `request` object if the `idGenerator` function parameter uses it.
```

File: test/session.test.js [hunk 8]
```diff
@@ -345,3 +345,114 @@ test('should update the expires property of the session using Session#touch() ev
...
+  fastify.register(fastifySession, {
+    secret: 'cNaoPYAwF60HZJzkcNaoPYAwF60HZJzk',
+    cookie: { secure: false, maxAge: 10000 },
+    idGenerator: (request) => {
+      if (request.session && request.session.returningVisitor) {
+        return `returningVisitor-${
+          new Date().getTime()
+        }-${
+          Math.random().toString().slice(2)
+        }`
```
Comment (Eomm) [inline test/session.test.js:361]: Could you split this statement into multiple ones?

File: test/session.test.js [hunk 9]
```diff
@@ -345,3 +345,114 @@ test('should update the expires property of the session using Session#touch() ev
...
+      }
+      return `custom-${
+        new Date().getTime()
+      }-${
+        Math.random().toString().slice(2)
+      }`
```
Comment (Eomm) [inline test/session.test.js:367]: Could you split this statement into multiple ones? (it is hard to read and there is some duplicated code

File: types/types.test-d.ts [hunk 10]
```diff
@@ -50,6 +50,10 @@ app.register(plugin, {
   secret: 'ABCDEFGHIJKLNMOPQRSTUVWXYZ012345',
   idGenerator: () => Date.now() + ''
 });
+app.register(plugin, {
+  secret: 'ABCDEFGHIJKLNMOPQRSTUVWXYZ012345',
+  idGenerator: (request) => request?.ip + '-' + Date.now()
```
Comment (darkgl0w) [inline types/types.test-d.ts:55]: Node.js 10 does not support optional chaining ^^

**Review by darkgl0w — APPROVED** [review]
lgtm

**Comment by Legolaszstudio** [comment]
I think everything requested is done, only waiting for
> One other way to type the `regenerate` method we can explore would be something like this:
> 
> ```ts
> // With this we can get the parameters of a function type in a tuple
> type Parameters<T> = T extends (...args: infer P) => any ? P : never;
> 
> // and by typing regenerate like this it will accept the same arguments (and types) that `idGenerator` accepts
> // NB: if we add another argument to `idGenerator` it will be added to `regenerate` too.
> regenerate(...args: Parameters<FastifySessionPlugin.Options['idGenerator']>): void;
> ```
> 
> cc @fastify/typescript WDYT ?
> 
> Edit: it seems that Typescript already has a `Parameters` utility defined like this:
> 
> ```ts
> type Parameters<T extends (...args: any) => any> = T extends (...args: infer P) => any ? P : never
> ```

**Review by mcollina — APPROVED** [review]
lgtm
